### PR TITLE
[GLUTEN-7267][CH]Support nested column pruning for `HiveTableScan`  json/parquet/orc format

### PR DESCRIPTION
**Pull request by KevinyhZou**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#7267)

## How was this patch tested?

BY UT

### CODE REVIEW

File: backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala [hunk 1]
```diff
@@ -1416,4 +1416,72 @@ class GlutenClickHouseHiveTableSuite
...
+    spark.sql(insert_sql_1)
+    spark.sql(insert_sql_2)
+    val select_sql_1 =
+      "select id, d1.c, d1.d[0].x, d2.d['m124'].y from %s where day = '2024-09-26' and hour = '12'"
+        .format(json_table_name)
+    val select_sql_2 =
+      "select id, d1.c, d1.d[0].x, d2.d['m124'].y from %s where day = '2024-09-26' and hour = '12'"
+        .format(pq_table_name)
+    withSQLConf(
+      ("spark.sql.hive.convertMetastoreParquet" -> "false"),
```
Comment (taiyang-li) [inline backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala:1460]: 这俩orc和parquet的开关在什么使用场景下是false呢

Comment (KevinyhZou) [reply backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala:1460]: 当需要使用hive parquet/orc serde 读取 table 时，而不是使用spark内置的parquet/orc reader读取时，这两个配置就需要被设置为false @taiyang-li

**Comment by KevinyhZou** [comment]
#### 性能测试
表schema：test_tbl (a STRING, b STRUCT<x1: STRING, x2: STRING, x3: STRING, x4: STRING, x5: STRING>)
测试sql： select count(b.x1) from test_tbl
数据量：1200W行
分别使用json/parquet/orc 三种测试存放数据，测试 该SQL查询的端到端耗时情况

优化前 平均耗时：
json格式： 16.52s
parquet耗时：2.02s
orc耗时：1.25s

优化后 平均耗时：
json格式：12.71s
parquet耗时： 0.63s
orc耗时：0.36s

File: shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala [hunk 2]
```diff
@@ -855,6 +858,13 @@ object GlutenConfig {
       .booleanConf
       .createWithDefault(true)
 
+  val COLUMNAR_HIVETABLESCAN_NESTED_COLUMN_PRUNING_ENABLED =
+    buildConf("spark.gluten.sql.columnar.hivetablescan.nested.column.pruning")
```
Comment (taiyang-li) [inline shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:862]: 跟其他config命名对齐，如spark.gluten.sql.columnar.enableVanillaVectorizedReaders

File: backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala [hunk 3]
```diff
@@ -1419,4 +1419,72 @@ class GlutenClickHouseHiveTableSuite
...
+      ("spark.sql.hive.convertMetastoreOrc" -> "false")) {
```
Comment (taiyang-li) [inline backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala:1464]: test中没有orc表，是否可去掉

File: backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala [hunk 4]
```diff
@@ -1419,4 +1419,72 @@ class GlutenClickHouseHiveTableSuite
     spark.sql("DROP TABLE test_tbl_7054")
   }
 
+  test("test hive table scan nested column pruning") {
+    val json_table_name = "test_tbl_7267_json"
```
Comment (taiyang-li) [inline backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala:1423]: 如果能写成 按照formats = ['parquet', 'orc']循环就更好了

File: gluten-substrait/src/main/scala/org/apache/spark/sql/hive/HiveTableScanNestedColumnPruning.scala [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+      }
+
+    // Construct the new projections of our Project by
+    // rewriting the original projections
+    val newProjects =
+      normalizedProjects.map(_.transformDown { case projectionOverSchema(expr) => expr }).map {
+        case expr: NamedExpression => expr
+      }
+
+    if (log.isDebugEnabled) {
```
Comment (taiyang-li) [inline gluten-substrait/src/main/scala/org/apache/spark/sql/hive/HiveTableScanNestedColumnPruning.scala:194]: remove.

File: gluten-substrait/src/main/scala/org/apache/spark/sql/hive/HiveTableScanNestedColumnPruning.scala [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+import org.apache.spark.sql.util.SchemaUtils._
+import org.apache.spark.util.Utils
+
+object HiveTableScanNestedColumnPruning extends Logging {
+  import org.apache.spark.sql.catalyst.expressions.SchemaPruning._
+
+  def supportNestedColumnPruning(projectExec: ProjectExec): Boolean = {
+    if (
+      GlutenConfig.getConf.enableColumnarHiveTableScanNestedColumnPruning
+      && BackendsApiManager.getSettings.supportHiveTableScanNestedColumnPruning()
```
Comment (taiyang-li) [inline gluten-substrait/src/main/scala/org/apache/spark/sql/hive/HiveTableScanNestedColumnPruning.scala:38]: Is it simpler if we decrease configs from 2 to 1?

File: backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala (hunk 4 above)
Comment (KevinyhZou) [reply backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala:1423]: 这里没有测试ORC format

**Comment by KevinyhZou** [comment]
@rui-mo take a look at this pr, whether velox backend need this feature ?

**Review by rui-mo — COMMENTED** [review]
@KevinyhZou In the Velox backend, it is some kind of reader's task. For the below case you mentioned, Velox Reader does not output the 'a2' and 'a3' columns. We are also working on some enhancements for the schema pruning in Velox, see https://github.com/facebookincubator/velox/pull/5962.

> Struct<a1 string, a2 string, a3 string>, when query s.a1 from table

**Comment by rui-mo** [comment]
I'm a little confused on the issue this PR is addressing. Are we supporting cases like those in the below suite?
https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruningSuite.scala

**Comment by KevinyhZou** [comment]
> I'm a little confused on the issue this PR is addressing. Are we supporting cases like those in the below suite? https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruningSuite.scala
Yes, we do this to only read the fields that we need from the complex type, like `array`, `map`, `struct`, which means to prune columns for the nested columns. we enabled this feature for `HiveTableScanTransformer`, and in gluten we use it to read json/hive text, and also can be used for parquet/orc (not the default way).

**Comment by rui-mo** [comment]
@KevinyhZou Thanks for your feedback. We have enabled 'GlutenParquetSchemaSuite' on Velox backend. The way we enable it is: Gluten passes user-specified schema to Velox, and the Velox reader handles the schema mismatch. For example, Gluten passes 'struct<a1>' as the output type of Velox scan node, and Velox reader handles the schema pruning internally. I wonder if CH needs this PR because the CH reader cannot handle schema pruning. Would you clarify? Thanks.

> Struct<a1 string, a2 string, a3 string>, when query s.a1 from table

**Comment by KevinyhZou** [comment]
OK,  I see. This feature is already enabled for clickhouse backend `FileSourceScan`,  which is the default way to read parquet/orc format data, we pass a prunned schema from `FileSourceScanTransformer` to ch parquet/orc reader.  And I think we would not change this way. thanks. @rui-mo

**Comment by taiyang-li** [comment]
Run Gluten Clickhouse CI on x86

**Review by taiyang-li — APPROVED** [review]
LGTM

**Comment by taiyang-li** [comment]
Run Gluten Clickhouse CI on x86

**Comment by zzcclp** [comment]
Run Gluten Clickhouse CI on x86

**Pull request merged**